frame/staking/src/lib.rs

```diff
@@ -761,6 +766,105 @@ impl<AccountId, Balance: HasCompact + Zero> UnappliedSlash<AccountId, Balance> {
+	// nominations based on `fn curve` and the nominator's balance.
+	type MaxNominations: Get<u32>;
+
+	/// Returns the voter's nomination quota within reasonable bounds [`min`, `max`], where `min`
+	/// is 1 and `max` is `Self::MaxNominations`.
+	fn get_quota(balance: Balance) -> u32 {
+		Self::curve(balance).clamp(1, Self::MaxNominations::get())
+	}
+
+	// Returns the voter's nomination quota based on its balance and a curve.
+	fn curve(balance: Balance) -> u32;
+}
+
+/// A nomination quota that allows up to MAX nominations for all validators.
+pub struct FixedNominationsQuota<const MAX: u32>;
+impl<Balance, const MAX: u32> NominationsQuota<Balance> for FixedNominationsQuota<MAX> {
+	type MaxNominations = ConstU32<MAX>;
+
+	fn curve(_: Balance) -> u32 {
+		MAX
+	}
+}
+
+/// A static tracker for the election data snapshot.
+///
+/// Computes the (SCALE) encoded byte length of a snapshot based on static rules, without any actual
+/// encoding.
+///
+/// ## Details
+///
+/// The snapshot has the form of `Vec<Voter>` where `Voter = (Account, u64, Vec<Account>)`. For
+/// each voter added to the snapshot, [`try_register_voter`] should be called, with the number
+/// of votes (length of the internal `Vec`).
+///
+/// Whilst doing this, [`size`] will track the entire size of the `Vec<Voter>`, except for the
+/// length prefix of the outer `Vec`. To get the final size at any point, use
+/// [`final_byte_size_of`].
+#[derive(Copy, Clone, Debug)]
+pub(crate) struct ElectionSizeTracker<AccountId> {
+	pub size: usize,
+	pub num_voters: usize,
+	_marker: sp_std::marker::PhantomData<AccountId>,
+}
+
+impl<AccountId> ElectionSizeTracker<AccountId>
+where
+	AccountId: IdentifierT,
+{
+	pub(crate) fn new() -> Self {
+		ElectionSizeTracker { size: 0, num_voters: 0, _marker: Default::default() }
+	}
+
+	/// Attempts to register a new voter with `votes` for a given election `bounds`. Returns an
+	/// error if the new size exceeds the capacity of the tracker.
+	pub(crate) fn try_register_voter(
+		&mut self,
+		votes: usize,
+		bounds: DataProviderBounds,
+	) -> Result<(), ()> {
+		let voters_size_after = self.size.saturating_add(Self::encoded_size(votes));
+
+		// the total encoded size takes into consideration the prefix of the outer vec, which
+		// contains all the voters.
+		let total_size_after = Self::final_byte_size_of(self.num_voters + 1, voters_size_after);
+
+		match bounds.size_exhausted(SizeBound(total_size_after as u32)) {
+			true => Err(()),
+			false => {
+				self.size = voters_size_after;
+				self.num_voters += 1;
+				Ok(())
+			},
+		}
+	}
+
+	/// Returns the size in MBs of the scale encoded structure that stores a `Voter` with a given
+	/// number of cast `votes`.
+	pub fn encoded_size(votes: usize) -> usize {
+		Self::length_prefix(votes)
+			.saturating_add(votes * sp_std::mem::size_of::<AccountId>())
+			.saturating_add(sp_std::mem::size_of::<frame_election_provider_support::VoteWeight>())
```

Comment:
I am doubting my original decision here to use `mem::size_of`. We are relying on the fact that `AccountId` is something that is only stored in the stack, and therefore: `sp_std::mem::size_of::<AccountId>() == some_account_id.encoded_size()`. In reality, i think we need to use `encoded_size` here. 

I suggest two things: 

* Play around with `encoded_size` and `size_of` for more insight into this. 
* We should double check the size bounding extensively before using it in production. Perhaps we can mark the API somehow as `unsafe` or similiar to make sure it is being used with caution. 

Recall that the end goal of this PR is to make sure the code path in EPM does not run out of allocation memory. Some of the PRs that fixed the OOM issue at https://polkadot.network/blog/a-polkadot-postmortem-24-05-2021/. 

I am pretty sure that with extnsive testing, especially with this PR, we can grow the size of nominator snapshot safely to something like `count: 40,000 | size: 14MB`, but we need to do really really extensive testing on it e2e before risking anything for it.

Reply:
All of this can be follow-ups. For now I only suggest making sure that there are some warnings in place when using `size` bound, either via compiler warning, or in the docs.

Reply:
I did some testing around the usage of `sp_std::mem::size_of`, [`Encoded::size_hint`](https://docs.rs/parity-scale-codec/latest/parity_scale_codec/trait.Encode.html#method.size_hint) and `encoded_size`. I think that the implementation is safer with `size_hint` and it achieves our goals: 1) no need to do extra allocations/encoding while iterating through the voters and 2) the resulting size of the tracker matches with the final `encoded_size` of the voters.

I also refactored all the code and tests related to the static size tracker to its own module and added more docs.

I'm now thinking about a way to add `try_state` checks that would allow us to extensively tests the size bounds and the static tracker. Once we have those implemented, we can run `follow-chain` using a runtime with different bounds and `try-state` enabled, so we can verify if the checks and configs work as expected.